File: py38/sort/insert.py
from utils import timer

from typing import List


@timer
def insert_sort(l: List):
    for i in range(1, len(l)):
        insert_num = l[i]

        j = i-1
        while j >= 0:
            if insert_num < l[j]:
                l[j+1] = l[j]
            else:
                break
            j -= 1
        j += 1

        l[j] = insert_num


if __name__ == "__main__":
    l = [1, 2, 4, 3]
    insert_sort(l)
    assert l == list(range(1, 5))

File: rs/src/sort/insert.rs
pub fn insert_sort<T: Ord + Copy>(l: &mut [T]) {
    for i in 1..l.len() {
        let insert_num = l[i];
        let mut j = i;

        while j > 0 {
            if insert_num < l[j-1] {
                l[j] = l[j-1];
            }
            else {
                break;
            }
            j -= 1;
        }

        l[j] = insert_num;
    }
}